cli/main.rs

```diff
@@ -822,3 +829,149 @@ fn wait_for_start(
     Ok(Some((unconfigured, args)))
   })
 }
+
+async fn auth_tunnel() -> Result<String, deno_core::anyhow::Error> {
+  let mut child = tokio::process::Command::new(env::current_exe()?)
+    .args(["deploy", "tunnel-login"])
+    .spawn()?;
+  let out = child.wait().await?;
+
+  if !out.success() {
+    deno_runtime::exit(1);
+  }
+
+  Ok(tools::deploy::get_token_entry()?.get_password()?)
+}
+
+#[allow(clippy::print_stderr)]
+async fn initialize_tunnel(
+  host: &str,
+  flags: &Flags,
+) -> Result<(), deno_core::anyhow::Error> {
+  use deno_lib::args::get_root_cert_store;
+
+  let factory = CliFactory::from_flags(Arc::new(flags.clone()));
+  let cli_options = factory.cli_options()?;
+  let deploy_config = cli_options.start_dir.to_deploy_config()?;
+  if deploy_config.is_none() {
+    let _ = tools::deploy::get_token_entry()?.delete_credential();
+  }
+
+  let token = if let Ok(token) = std::env::var("DENO_UNSTABLE_TUNNEL_TOKEN") {
+    token
+  } else {
+    match tools::deploy::get_token_entry()?.get_password() {
+      Ok(token) => token,
+      Err(keyring::Error::NoEntry) => auth_tunnel().await?,
+      Err(e) => {
+        init_logging(None, None);
+        exit_for_error(e.into());
+      }
+    }
+  };
+
+  let factory = CliFactory::from_flags(Arc::new(flags.clone()));
```

Comment:
It looks like this code is creating a factory twice? (See line 853) That said, why is it even creating a factory? That should be done at a very high level. I think code using a tunnel will do config discovery 3 times on startup.

Reply:
in `auth_tunnel`, so a subprocess